File: app/src/main/java/com/android/virgilsecurity/virgilback4app/utils/UsernameInputFilter.java
package com.android.virgilsecurity.virgilback4app.utils;

import android.text.InputFilter;
import android.text.Spanned;

public class UsernameInputFilter implements InputFilter {

    public CharSequence filter(CharSequence source, int start, int end,
                               Spanned dest, int dstart, int dend) {

        final String constraint = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,-()/='+:?!%&*<>;{}@#_";

        StringBuilder builder = new StringBuilder();
        for (int i = start; i < end; i++) {
            char c = source.charAt(i);
            if (constraint.contains(String.valueOf(c))) {
                builder.append(c);
            }
        }

        boolean allCharactersValid = (builder.length() == end - start);
        return allCharactersValid ? null : builder.toString();
    }
}

File: app/src/main/java/com/android/virgilsecurity/virgilback4app/util/UsernameInputFilter.kt
package com.android.virgilsecurity.virgilback4app.util

import android.text.InputFilter
import android.text.Spanned

class UsernameInputFilter : InputFilter {

    override fun filter(source: CharSequence, start: Int, end: Int,
                        dest: Spanned, dstart: Int, dend: Int): CharSequence? {

        val constraint = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,-()/='+:?!%&*<>;{}@#_"

        val builder = StringBuilder()
        for (i in start until end) {
            val c = source[i]
            if (constraint.contains(c.toString())) {
                builder.append(c)
            }
        }

        val allCharactersValid = builder.length == end - start
        return if (allCharactersValid) null else builder.toString()
    }
}